### ISSUE TITLE
Full Stack Deep Learning - Course Note

### ISSUE THREAD

**Issue by amyxie361** [amyxie361/Blog#32]
2021 Winter, Following [FSDL](https://fullstackdeeplearning.com/), [Youtube](https://www.youtube.com/channel/UCVchfoB65aVtQiDITbGq2LQ)
Some notes and inspirations.

**Comment by amyxie361** [comment]
## Lecture 2B: ConvNets.

### Classification
**History of ImageNet**
2012, AlexNet 83.6. Contribution: ReLU, Dropout, Data Augmentation(flip, scale, ...)
![image](https://user-images.githubusercontent.com/26669286/107288022-cc178500-6a30-11eb-9902-c783c6de5f4e.png)

2013, ZFNet 88.3. Contribution: smaller conv1, higher channel conv3,4,5; DeConv Visiualization

2014, VGG 92.7. Contribution: all conv 3x3, deeper, max pooling
![image](https://user-images.githubusercontent.com/26669286/107288350-4f38db00-6a31-11eb-909f-1906654789d1.png)

2014, GoogleNet, 93.3, Contribution: Inception Modules, different structure for different channels, Early exit
![image](https://user-images.githubusercontent.com/26669286/107288563-a474ec80-6a31-11eb-89c2-225375fee0c1.png)

2015, ResNet, 96.4, Contribution: add residual connection to avoid vanish gradient, strides instead of max pooling
![image](https://user-images.githubusercontent.com/26669286/107288743-f0279600-6a31-11eb-9805-1dd90af0cddb.png)
- Dense Net: add more connections
- ResNeXT
2017 SENet, 97.7, Contribution: Add Global pooling 

2017 Squeeze net, Contribution: add conv1x1 to reduce channels.
![image](https://user-images.githubusercontent.com/26669286/107310761-10b81600-6a5b-11eb-9294-259428cbe860.png)

![image](https://user-images.githubusercontent.com/26669286/107310886-4bba4980-6a5b-11eb-83dc-92c10e90d9ab.png)

Top 1 Acc: 80

### Other tasks:

Localization: Classification + find the bounding box
Detection: find all kinds of objects in the picture
Segmentation: label each pixel to it's class

2013 Overfeat: move the bounding box + classification
YOLO/SSD
R-CNN, Faster R-CNN, 
Region Proposal Network
Mask R-CNN
Fully conv Net: Encode + Decode(How to upsampling)

3D: Mesh R-CNN (Dataset: shapeNet)
Face Landmarks(Dataset: FAWL)
Pose Estimation(COCO)

Adversarial Attacks(White box / Black box: can access the original model or not)
Defence:
- adding adversarial examples to training: not work well
- smooth loss function (Defence distillation)

Style Transfer

GANS

**Comment by amyxie361** [comment]
## Lec 3: RNNs

### Sequence problems:
Time series forecasting
Sentiment classification
Translation
Speech recognition and generation
Text or music generation
Image captioning
Question Answering

**Why not FFN?**
Cannot deal with arbitrary length

RNN: yt, ht = f(x, ht-1)
LSTM/GRUs

### Key questions for Applications:
- what problem are they trying to solve
- what model architecture
- what loss function
- what dataset trained on
- how do they train
- are there tricks during inference

LSTM(underfit)-> stack LSTMs(hard to train)->ResLSTM
Too much information at bottle neck(between encoder and decoder)
-> Attention: give entire history to the decoder
Visualization: connection/heat map

Future information are also important: Bi-directional LSTM

### CTC loss
Input is not in a fixed units (hand writing/speech)
still input with fixed length window, after prediction, merge repeated outputs.
+ \epsilon to distinguish between real duplicated tokens (Similar as in BIO)

### Pros and Cons
**Pros**:
- can model arbitrary sequence problems
- Many successes
**Cons**
- not as parallelizable as CNN/FC
- slower
- finicky to train

### non RNNs:
WaveNet - ConvNets
